src/help.cpp

```diff
@@ -157,136 +115,216 @@ std::string help::get_note_colors()
-
-                    const auto get_w_help_border = [&]() {
-                        init_windows( ui );
-                        return w_help_border;
-                    };
-
-                    scrollable_text( get_w_help_border, _( "Help" ),
-                                     std::accumulate( i18n_help_texts.begin() + 1, i18n_help_texts.end(),
-                                                      i18n_help_texts.front(),
-                    []( std::string lhs, const std::string & rhs ) {
-                        return std::move( lhs ) + "\n\n" + rhs;
-                    } ) );
-
-                    ui.on_screen_resize( init_windows );
+void help_window::draw_category_option( const int &option, const help_category &category )
+{
+    std::string cat_name;
+    auto hotkey_it = hotkeys.find( option );
+    if( hotkey_it != hotkeys.end() ) {
+        cat_name = hotkey_it->second.short_description();
+        cat_name += ": ";
+    }
+    cat_name += category.name.translated();
+    if( data.read_categories.find( option ) != data.read_categories.end() ) {
+        ImGui::PushStyleColor( ImGuiCol_Text, c_light_gray );
+        ImGui::Selectable( remove_color_tags( cat_name ).c_str() );
+        ImGui::PopStyleColor();
+    } else {
+        ImGui::Selectable( remove_color_tags( cat_name ).c_str() );
+    }
+    if( ImGui::IsItemHovered() ) {
+        selected_option = option;
+    }
+}
+
+std::string help_window::seperator( int length, char c )
+{
+    std::string ret = "<color_light_blue>";
+    ret += std::string( length, c );
+    ret += "</color>";
+    return ret;
+}
+
+void help_window::format_title( const std::string translated_category_name )
+{
+    cataimgui::PushMonoFont();
+    if( get_option<bool>( "SCREEN_READER_MODE" ) ) {
+        cataimgui::draw_colored_text( translated_category_name );
+        ImGui::NewLine();
+        return;
+    }
+    std::string row3 = "<color_light_blue>║</color> ";
+    row3 += translated_category_name;
+    row3 += " <color_light_blue>║</color>";
+    std::string div = seperator( remove_color_tags( row3 ).length() - 4, '=' );
+    cataimgui::draw_colored_text( div );
+    cataimgui::draw_colored_text( row3 );
+    cataimgui::draw_colored_text( div );
+    cataimgui::draw_colored_text( seperator( TERMX, '_' ) );
+    ImGui::NewLine();
+    ImGui::PopFont();
+}
+
+void help_window::draw_category()
+{
+    const help_category &cat = data.help_categories[loaded_option];
+    format_title( cat.name.translated() );
+    if( ImGui::BeginTable( "HELP_PARAGRAPHS", 1,
+                           ImGuiTableFlags_ScrollY ) ) {
+        ImGui::TableNextRow();
+        ImGui::TableNextColumn();
+        for( const translation &paragraph : cat.paragraphs ) {
+            std::string translated_paragraph = paragraph.translated();
+            parse_tags_help_window( translated_paragraph );
+            cataimgui::draw_colored_text( translated_paragraph, wrap_width );
```

Comment:
If you use `cataimgui::TextColoredParagraph` instead, with no `wrap_width`, then you’ll get pixel–correct wrapping even on text containing color tags. `cataimgui::draw_colored_text` will be retired eventually.

Reply:
That's what I were using to begin with it was just failing to interpret this correctly so I switched and then forgot about it, didn't appreciate draw_colored_text was temporary, should I just be using TextColoredParagraph everywhere?
![image](https://github.com/user-attachments/assets/efd45a42-04a3-4488-bde9-9091b164b289)
I'll take a look to see if I can fix that easily or I'll open an issue otherwise

Reply:
It messes up a bit when the “paragraph” that you give it contains embedded newline characters. Currently the best way to avoid the problem is to split paragraphs into lines and then call TextColoredParagraph on each one. 

Also, it doesn’t handle incorrect tags very gracefully, so maybe you just need to fix the JSON for those.

Reply:
> Also, it doesn’t handle incorrect tags very gracefully, so maybe you just need to fix the JSON for those.

The JSON is fine that's showing the 4 rotations of buildings, it's failing to recognise the </color> bc of the preceding <
The current window and draw_colored_text handles it correctly. I can probably just stick a space between them in the JSON for this one instance like but it shouldn't be needed.

Reply:
Oooh, I see now. I’ll see if I can improve the code to handle it better.

Reply:
> Currently the best way to avoid the problem is to split paragraphs into lines and then call TextColoredParagraph on each one.

```c++
void help_window::dir_grid()
{
    static const std::array<std::vector<input_event>, 9> movearray = {{
            keys_bound_to( ACTION_MOVE_FORTH_LEFT, 0 ), keys_bound_to( ACTION_MOVE_FORTH, 0 ), keys_bound_to( ACTION_MOVE_FORTH_RIGHT, 0 ),
            keys_bound_to( ACTION_MOVE_LEFT, 0 ),  keys_bound_to( ACTION_PAUSE, 0 ),  keys_bound_to( ACTION_MOVE_RIGHT, 0 ),
            keys_bound_to( ACTION_MOVE_BACK_LEFT, 0 ), keys_bound_to( ACTION_MOVE_BACK, 0 ), keys_bound_to( ACTION_MOVE_BACK_RIGHT, 0 )
        }
    };

    static const std::array<std::string, 9> movearray_hjkl = {{
            movearray[0].size() > 0 ? string_format( "<color_light_blue>%s</color>", movearray[0][0].short_description() ) : "<color_red>?</color>",
            movearray[1].size() > 0 ? string_format( "<color_light_blue>%s</color>", movearray[1][0].short_description() ) : "<color_red>?</color>",
            movearray[2].size() > 0 ? string_format( "<color_light_blue>%s</color>", movearray[2][0].short_description() ) : "<color_red>?</color>",
            movearray[3].size() > 0 ? string_format( "<color_light_blue>%s</color>", movearray[3][0].short_description() ) : "<color_red>?</color>",
            movearray[4].size() > 0 ? string_format( "<color_light_blue>%s</color>", movearray[4][0].short_description() ) : "<color_red>?</color>",
            movearray[5].size() > 0 ? string_format( "<color_light_blue>%s</color>", movearray[5][0].short_description() ) : "<color_red>?</color>",
            movearray[6].size() > 0 ? string_format( "<color_light_blue>%s</color>", movearray[6][0].short_description() ) : "<color_red>?</color>",
            movearray[7].size() > 0 ? string_format( "<color_light_blue>%s</color>", movearray[7][0].short_description() ) : "<color_red>?</color>",
            movearray[8].size() > 0 ? string_format( "<color_light_blue>%s</color>", movearray[8][0].short_description() ) : "<color_red>?</color>"
        }
    };

    static const std::array<std::string, 9> movearray_numpad = {{
            movearray[0].size() > 1 ? string_format( "<color_light_blue>%s</color>", movearray[0][1].short_description() ) : "<color_red>?</color>",
            movearray[1].size() > 1 ? string_format( "<color_light_blue>%s</color>", movearray[1][1].short_description() ) : "<color_red>?</color>",
            movearray[2].size() > 1 ? string_format( "<color_light_blue>%s</color>", movearray[2][1].short_description() ) : "<color_red>?</color>",
            movearray[3].size() > 1 ? string_format( "<color_light_blue>%s</color>", movearray[3][1].short_description() ) : "<color_red>?</color>",
            movearray[4].size() > 1 ? string_format( "<color_light_blue>%s</color>", movearray[4][1].short_description() ) : "<color_red>?</color>",
            movearray[5].size() > 1 ? string_format( "<color_light_blue>%s</color>", movearray[5][1].short_description() ) : "<color_red>?</color>",
            movearray[6].size() > 1 ? string_format( "<color_light_blue>%s</color>", movearray[6][1].short_description() ) : "<color_red>?</color>",
            movearray[7].size() > 1 ? string_format( "<color_light_blue>%s</color>", movearray[7][1].short_description() ) : "<color_red>?</color>",
            movearray[8].size() > 1 ? string_format( "<color_light_blue>%s</color>", movearray[8][1].short_description() ) : "<color_red>?</color>"
        }
    };

    cataimgui::TextColoredParagraph( c_white, string_format( "%s  %s  %s   %s  %s  %s",
                                     movearray_numpad[0], movearray_numpad[1], movearray_numpad[2], movearray_hjkl[0], movearray_hjkl[1],
                                     movearray_hjkl[2] ) );
    ImGui::NewLine();
    cataimgui::PushMonoFont();
    cataimgui::TextColoredParagraph( c_white, " \\ | /     \\ | /" );
    ImGui::NewLine();
    cataimgui::TextColoredParagraph( c_white, "  \\|/       \\|/" );
    ImGui::NewLine();
    ImGui::PopFont();
    cataimgui::TextColoredParagraph( c_white, movearray_numpad[3] );
    cataimgui::PushMonoFont();
    cataimgui::TextColoredParagraph( c_white, "--" );
    ImGui::PopFont();
    cataimgui::TextColoredParagraph( c_white, movearray_numpad[4] );
    cataimgui::PushMonoFont();
    cataimgui::TextColoredParagraph( c_white, "--" );
    ImGui::PopFont();
    cataimgui::TextColoredParagraph( c_white, movearray_numpad[5] );
    cataimgui::PushMonoFont();
    cataimgui::TextColoredParagraph( c_white, "   " );
    ImGui::PopFont();
    cataimgui::TextColoredParagraph( c_white, movearray_hjkl[3] );
    cataimgui::PushMonoFont();
    cataimgui::TextColoredParagraph( c_white, "--" );
    ImGui::PopFont();
    cataimgui::TextColoredParagraph( c_white, movearray_hjkl[4] );
    cataimgui::PushMonoFont();
    cataimgui::TextColoredParagraph( c_white, "--" );
    ImGui::PopFont();
    cataimgui::TextColoredParagraph( c_white, movearray_hjkl[5] );
    ImGui::NewLine();
    cataimgui::PushMonoFont();
    cataimgui::TextColoredParagraph( c_white, "  /|\\       /|\\" );
    ImGui::NewLine();
    cataimgui::TextColoredParagraph( c_white, " / | \\     / | \\" );
    ImGui::NewLine();
    ImGui::PopFont();
    cataimgui::TextColoredParagraph( c_white, string_format( "%s  %s  %s   %s  %s  %s",
                                     movearray_numpad[6], movearray_numpad[7], movearray_numpad[8], movearray_hjkl[6], movearray_hjkl[7],
                                     movearray_hjkl[8] ) );
    ImGui::NewLine();
}
```
I mean this doesn't seem very sane >_> (I appreciate I could probably build the latter two arrays iteratively but it's still gross) Ig I could make it return a static const array of lines of strings instead...? Pretty annoying that it can't just handle \n tho

Reply:
> Oooh, I see now. I’ll see if I can improve the code to handle it better.

Adding spaces between the < and </color> also doesn't fix this

Reply:
> Pretty annoying that it can't just handle \n tho

I know, I really gotta get in there and fix that bug. I’ll make it a goal for tomorrow.

> Adding spaces between the < and also doesn't fix this

If you look at text.cpp line 297, you’ll see that it finds the first `<` and then scans until it finds a `>`.  It calls `get_color_from_tag` to get a color, but in this case it finds `<</whatever>` as the tag which doesn’t match anything. Nobody really thought this through, I think.

Reply:
BTW, we could consider using images here, if you want. Of course ASCII art is not a bad stylistic choice either.

Reply:
I think I'm pretty pro ASCII stuff in general but in this case it just seems more awkward to do anything else, it can't be a single image bc it's showing the current binds which I think is important and trying to piece together images around the binds would probably be just as bad, it looks clear in game as is

It could be done with something without lines like
```
NW N NE    7 8 9    y k u
W  O  E    4 5 6    h . l
SW S SE    1 2 3    b j n
```
(still using the current binds) but that feels less useful

Reply:
ImGui is perfectly happy to overlap things; it could be a single image with text overlaid. Imagine a nice little compass rose with the points labeled by keybinds. We can measure the exact bounding box of each keybind, and position them correctly so that they always line up right, don’t overlap, don’t spill out of the window, etc. Instead of using spaces and `ImGui::NewLine()` to roughly position things, we can use `ImGui::SetCursorPos()` to put things at exact coordinates.

Or there are unicode characters:
```
↖↑↗
←❂→
↙↓↘
```